Recent Topics: Stage 1

aero31aero opened the pull request:
This starts work on a new feature to show list of recent topics that a user participated in, and take a quick look at the ones where someone has replied to him.

This is the module's structure so far:

1. `recent_topics.process_all_messages()` - to be called once when we have loaded all messages.
2. `recent_topics.process(message)` - to be called each time we get a message from server.
3.  `recent_topics.get()` - returns all topics where we have sent a message.
4. `recent_topics.get_updated()` - returns all topics where someone has replied after us.

@showell @timabbott any initial feedback?

Original issue: #6605 

- [x] First Draft
- [x] Node Tests
- [x] Call this module in all needed message events
- [ ] Deploy on chat.zulip.org
- [ ] Next PR: Add UI elements to show list

showell commented on static/js/recent_topics.js:
```diff
@@ -0,0 +1,72 @@
+
+var exports = {};
+
+var topics = [];
```
You should consider using our `Dict` class here, which can handle case-insensitive keys and has a simple set/get/has API.

showell commented:
I had a small comment, but this seems reasonable so far.

aero31aero commented:
Updated and added tests. There were two strategies I had in mind:

1. Merge it in close to its present state and debug on a server with lots of messages.
2. Integrate with the main code before merging.

What do you suggest?

aero31aero commented:
@showell this is ready for another review. I'm not sure if I added the calls to this module in he correct place, can you please advice?

showell commented:
Looks great so far!  I assume this is still wip, though.

aero31aero commented:
Rebased and resolved conflicts.

showell commented on static/js/recent_topics.js:
```diff
@@ -0,0 +1,79 @@
+};
+
+exports.process_message = function (msg) {
+    var is_ours = msg.sender_id === page_params.user_id;
```
Can you change this to `is_ours = people.is_my_user_id(msg.sender_id)`?

showell commented on static/js/recent_topics.js:
```diff
@@ -0,0 +1,79 @@
+    var is_ours = msg.sender_id === page_params.user_id;
+    var is_relevant = is_ours && msg.type === 'stream';
+    var key = msg.stream_id + ':' + msg.subject;
+    var topic = topics.has(key) ? topics.get(key) : undefined;
```
You can simplify this to `var topic = topics.get(key)`, which will return `undefined` if the key doesn't exist.

showell commented on static/js/recent_topics.js:
```diff
@@ -0,0 +1,79 @@
+    }
+    if (!topic) {
+        topics.set(key, {
+            our_last_msg: msg,
```
I would consider using message ids here, instead of `msg`, as the way to remember recent messages.  This way we don't have to worry about the `msg` data getting stale due to message edits, etc.  (Well, we still need to think about message edits, but in principle we can let the downstream code call something like `message_store.get(msg_id)` when we need actual content.

aero31aero replied:
I was using the complete message object for easily comparing the timestamp. In any case, I would have to make a new object to store both the ids and the timestamps for comparison.

I'll make sure that the downstream code calls `message_store.get()`, and thus, changing it to return message ids instead.

showell commented on frontend_tests/node_tests/recent_topics.js:
```diff
@@ -0,0 +1,208 @@
+
+    messages[6] = {
+        stream_id: stream1,
+        timestamp: _.uniqueId(),
```
I'd consider just using hard coded timestamps in the test, so it's more obviously deterministic.

aero31aero commented:
Internally, I'm using three properties in the message objects: id, timestamp and sender, so I feel it doesn't make sense for me to not store the message objects completely. I've changed the external interface as follows:

```js
var topics = recent_topics.get_relevant();
```

returns a dict of objects like: 

```
key: "1:Denmark1",
value: {
    last_msg_id: 59,
    our_last_msg_id: 36,
    read: true,
}
```

aero31aero commented:
Closing this, @amanagr has built on top of this already.
